### ISSUE TITLE
"Service 'kube-httpcache' has no endpoints" when installing via Helm Chart

### ISSUE THREAD

**Issue by mahgo** [mittwald/kube-httpcache#79]
**Describe the bug**
I'm installing kube-httpcache via the Helm chart: `helm install kube-httpcache ./chart`. However it looks like it has issues starting, showing the error `service 'kube-httpcache' has no endpoints`:

`kubectl logs kube-httpcache-0`
```
I0601 06:34:17.607402       1 main.go:31] running kube-httpcache with following options: {Kubernetes:{Config: RetryBackoffString:30s RetryBackoff:30s} Frontend:{Address:0.0.0.0 Port:80 Watch:true Namespace:default Service:kube-httpcache PortName:http} Backend:{Watch:true Namespace:default Service:backend-service Port: PortName:http} Signaller:{Enable:true Address:0.0.0.0 Port:8090 WorkersCount:1 MaxRetries:5 RetryBackoffString:30s RetryBackoff:30s} Admin:{Address:0.0.0.0 Port:6083} Varnish:{SecretFile:/etc/varnish/k8s-secret/secret Storage:malloc,128M AdditionalParameters: VCLTemplate:/etc/varnish/tmpl/default.vcl.tmpl VCLTemplatePoll:false} Readiness:{Enable:true Address:0.0.0.0:9102}}
I0601 06:34:17.607529       1 main.go:38] using in-cluster configuration
I0601 06:34:17.611383       1 run.go:15] waiting for initial configuration before starting Varnish
W0601 06:34:17.635837       1 endpoints_watch.go:50] service 'kube-httpcache' has no endpoints
```

While my cluster does have RBAC enabled, it looks like the Helm chart handles setting up the ServiceAccount etc.

**To Reproduce**
`helm install kube-httpcache ./chart --set replicaCount=2`
`kubectl logs kube-httpcache-0`

**Expected behavior**
Should start without issues. 

**Environment:**
 - Kubernetes version: 
 ```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.7", GitCommit:"1dd5338295409edcfff11505e7bb246f0d325d15", GitTreeState:"clean", BuildDate:"2021-01-13T13:23:52Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.14", GitCommit:"4d62230f10af41cec2851b5b113874f6d2098297", GitTreeState:"clean", BuildDate:"2021-03-22T23:01:55Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```
 - kube-httpcache version: latest on master branch

**Configuration**
For examples sake I'm using the default vcl, I have the exact same issue when using my own configuration.

**Comment by ariful-khan** [comment]
**Describe the bug**
I am having the same issue with helm chart.

`I0603 10:22:50.083639       1 main.go:31] running kube-httpcache with following options: {Kubernetes:{Config: RetryBackoffString:30s RetryBackoff:30s} Frontend:{Address:0.0.0.0 Port:80 Watch:true Namespace:feature3 Service:varnish PortName:http} Backend:{Watch:true Namespace:feature3 Service:backend-service Port: PortName:http} Signaller:{Enable:true Address:0.0.0.0 Port:8090 WorkersCount:1 MaxRetries:5 RetryBackoffString:30s RetryBackoff:30s} Admin:{Address:0.0.0.0 Port:6083} Varnish:{SecretFile:/etc/varnish/k8s-secret/secret Storage:malloc,128M AdditionalParameters: VCLTemplate:/etc/varnish/tmpl/default.vcl.tmpl VCLTemplatePoll:false} Readiness:{Enable:true Address:0.0.0.0:9102}}
I0603 10:22:50.083750       1 main.go:38] using in-cluster configuration
I0603 10:22:50.084857       1 run.go:15] waiting for initial configuration before starting Varnish
W0603 10:22:50.095852       1 endpoints_watch.go:50] service 'varnish' has no endpoints
`

**Configuration**
Here is my config, created by following https://varnish-cache.org/docs/6.4/reference/vcl.html#local-server-remote-and-client

`  vcl 4.0;

  import std;
  import directors;

  // ".Frontends" is a slice that contains all known Varnish instances
  // (as selected by the service specified by -frontend-service).
  // The backend name needs to be the Pod name, since this value is compared
  // to the server identity ("server.identity" [1]) later.
  //
  //   [1]: https://varnish-cache.org/docs/6.4/reference/vcl.html#local-server-remote-and-client

  backend frontend0 {
        .host = "varnish-0.varnish.${NAMESPACE}.svc.cluster.local";
        .port = "80";
  }

  backend frontend1 {
        .host = "varnish-1.varnish.${NAMESPACE}.svc.cluster.local";
        .port = "80";
  }


  backend backend0 {
        .host = "app-0.app.${NAMESPACE}.svc.cluster.local";
        .port = "80";
            .first_byte_timeout = 60s;
            .connect_timeout = 3.5s;
            .between_bytes_timeout = 60s;
  }

  backend backend1 {
      .host = "app-1.app.${NAMESPACE}.svc.cluster.local";
      .port = "80";
          .first_byte_timeout = 60s;
          .connect_timeout = 3.5s;
          .between_bytes_timeout = 60s;
  }

  sub vcl_init {
      new cluster = directors.hash();

      cluster.add_backend(frontend0, 1);
      cluster.add_backend(frontend1, 1);

      new lb = directors.round_robin();

      lb.add_backend(backend0);
      lb.add_backend(backend1);
  }


  // Called at the beginning of a request, after the complete request has been received
  sub vcl_recv {
      # Routing logic. Pass a request to an appropriate Varnish node.
      # See https://info.varnish-software.com/blog/creating-self-routing-varnish-cluster for more info.

      set req.backend_hint = cluster.backend(req.url);
      set req.http.X-shard = req.backend_hint;

      if (req.http.X-shard == server.identity) {
        set req.backend_hint = lb.backend();
      } else {
        return(pass);
      }`

......
}

................

**Comment by mahgo** [comment]
@martin-helmich Do you know why this might be?

**Comment by martin-helmich** [comment]
> @martin-helmich Do you know why this might be?

Sorry, I haven't had time to investigate this, yet.

Could you maybe post the output of `kubectl describe pod kube-httpcache-0` (to see the actual flags `kube-httpcache` was started with) and the output of `kubectl get endpoints` (to see which services exist with which endpoints)?

**Comment by mahgo** [comment]
Sure, here they are:
```
Name:         kube-httpcache-0
Namespace:    default
Priority:     0
Node:         aks-agentpool-34153171-vmss000000/10.1.0.4
Start Time:   Wed, 02 Jun 2021 14:40:51 +1000
Labels:       app.kubernetes.io/instance=kube-httpcache
              app.kubernetes.io/name=kube-httpcache
              controller-revision-hash=kube-httpcache-6b9c4cb458
              statefulset.kubernetes.io/pod-name=kube-httpcache-0
Annotations:  <none>
Status:       Running
IP:           10.1.0.22
IPs:
  IP:           10.1.0.22
Controlled By:  StatefulSet/kube-httpcache
Containers:
  kube-httpcache:
    Container ID:  docker://d8f49b896e3a681e5fefa2733932f311ff9f9049c91bf0e7ff0c853092e14d76
    Image:         quay.io/mittwald/kube-httpcache:stable
    Image ID:      docker-pullable://quay.io/mittwald/kube-httpcache@sha256:073896f44f0d47cc14fbfb6b2a6e541de9bc0e64bf8097bbc76d372aa3effe27
    Port:          <none>
    Host Port:     <none>
    Args:
      -admin-addr=0.0.0.0
      -admin-port=6083
      -signaller-enable
      -signaller-port=8090
      -frontend-watch
      -frontend-namespace=$(NAMESPACE)
      -frontend-service=kube-httpcache
      -backend-watch
      -backend-namespace=$(NAMESPACE)
      -backend-service=backend-service
      -varnish-secret-file=/etc/varnish/k8s-secret/secret
      -varnish-vcl-template=/etc/varnish/tmpl/default.vcl.tmpl
      -varnish-storage=malloc,128M
    State:          Running
      Started:      Wed, 02 Jun 2021 14:42:33 +1000
    Ready:          True
    Restart Count:  0
    Environment:
      NAMESPACE:  default (v1:metadata.namespace)
    Mounts:
      /etc/varnish/k8s-secret from secret (rw)
      /etc/varnish/tmpl from template (rw)
      /var/lib/varnish from var (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-httpcache-token-g4vsv (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  template:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      kube-httpcache
    Optional:  false
  secret:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  kube-httpcache
    Optional:    false
  var:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
  kube-httpcache-token-g4vsv:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  kube-httpcache-token-g4vsv
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:          <none>
```

```
NAME                                               ENDPOINTS                                                AGE
kube-httpcache                                     10.1.0.16:8090,10.1.0.22:8090,10.1.0.16:80 + 1 more...   6d
kubernetes                                         20.193.8.129:443                                         18d
magento                                            10.1.0.17:8443,10.1.0.17:8080                            3h55m
magento-elasticsearch-coordinating-only            10.1.0.29:9200,10.1.0.29:9300                            3h55m
magento-elasticsearch-data                         10.1.0.36:9200,10.1.0.36:9300                            3h55m
magento-elasticsearch-master                       10.1.0.10:9200,10.1.0.10:9300                            3h55m
nginx-ingress-ingress-nginx-controller             10.1.0.5:443,10.1.0.5:80                                 3d23h
nginx-ingress-ingress-nginx-controller-admission   10.1.0.5:8443                                            3d23h
```

**Comment by mahgo** [comment]
Also here's a describe of the kube-httpcache endpoint, in case it is useful:
```
Name:         kube-httpcache
Namespace:    default
Labels:       app.kubernetes.io/instance=kube-httpcache
              app.kubernetes.io/managed-by=Helm
              app.kubernetes.io/name=kube-httpcache
              app.kubernetes.io/version=0.0.1
              helm.sh/chart=kube-httpcache-0.0.1
              test.test=test
Annotations:  endpoints.kubernetes.io/last-change-trigger-time: 2021-06-02T04:47:27Z
Subsets:
  Addresses:          10.1.0.16,10.1.0.22
  NotReadyAddresses:  <none>
  Ports:
    Name       Port  Protocol
    ----       ----  --------
    signaller  8090  TCP
    http       80    TCP

Events:  <none>
```

**Comment by williamroboly** [comment]
Sadly, I to am seeing this similar error. Any ideas on what this could be?

**Comment by mahgo** [comment]
@martin-helmich Have you had a chance to replicate this yet?

**Comment by williamroboly** [comment]
All I know is that once I've setup the stateful service and hook it all up, once it starts, the error message keeps logging.

```
I0617 03:20:30.367790       1 main.go:31] running kube-httpcache with following options: {Kubernetes:{Config: RetryBackoffString:30s RetryBackoff:30s} Frontend:{Address:0.0.0.0 Port:80 Watch:true Namespace:ciodrcoe-dev Service:frontend-service PortName:http} Backend:{Watch:true Namespace:ciodrcoe-dev Service:backend-service Port: PortName:http} Signaller:{Enable:true Address:0.0.0.0 Port:8090 WorkersCount:1 MaxRetries:5 RetryBackoffString:30s RetryBackoff:30s} Admin:{Address:0.0.0.0 Port:6083} Varnish:{SecretFile:/etc/varnish/k8s-secret/secret Storage:malloc,128M AdditionalParameters: VCLTemplate:/etc/varnish/tmpl/default.vcl.tmpl VCLTemplatePoll:false WorkingDir:} Readiness:{Enable:true Address:0.0.0.0:9102}}
I0617 03:20:30.367941       1 main.go:38] using in-cluster configuration
I0617 03:20:30.372200       1 run.go:15] waiting for initial configuration before starting Varnish
E0617 03:20:30.382634       1 endpoints_watch.go:29] error while establishing watch: unknown (get endpoints)
I0617 03:20:30.382651       1 endpoints_watch.go:30] retrying after 30s
```

I'm wondering if in the namespace, it is failing to find any active services to use. Though I have quite a few. Not exactly certain where to go from here.

**Comment by williamroboly** [comment]
BTW, thank you for this repo and the work you've done here. Its actually been super helpful!

**Comment by mahgo** [comment]
@martin-helmich Don't suppose you've had a chance to look at this yet have you?

**Comment by d0d00n** [comment]
I just deployed this template so I think I can help.

By default, the arg backend-service is set to the generic name backend-service. But you are to replace that with your own backend so if you wanted a service named magento, you'd go to your kube-httpcache statefulset and change

            - '-backend-service=backend-service'
          
to 

            - '-backend-service=magento'

or you can create a service called backend-service, I suppose.

Hope that fixes your issue!

**Comment by mahgo** [comment]
Thanks @d0d00n, this seems to be it!

**Comment by williamroboly** [comment]
Hey Don, thanks for looking into this.

The issue actually wasn’t getting it deployed, but rather ensuring the working varnish didn’t throw errors AND that the HTTPS actually worked.

I got it to consume and spit out a HTTP site fine, but the path to SSL was torqued.

I’m abandoning the solution for now as using it would require that we put in place either an HTTPS cert solution OR get around the bugs that I encountered. Regardless, after listening to what Filipe said, I was convince that using routes with our varnish solution will work in the interim.

If you were able to get everything sorted, and it was a simple configuration issue, I would be happy to walk through it all with you to get a better grasp on it.

Maybe we can talk this week?

-w
